cartridges/int_algolia/cartridge/scripts/algolia/helper/jobHelper.js

```diff
@@ -475,83 +488,128 @@ function _updateOrAddValue(objectsArray, key, value) {
- *          [...]
- *     },
- *     {
- *         'productID4': true,
- *          [...]
- *     },
- *    [...]
- * ]
- * `changedProducts` is an array of objects with each object containing at most 2000 key-value pairs.
- * _updateOrAddValue() makes sure that the keys are unique across the whole structure.
- * It was necessary to split the data up into multiple objects due to the SFCC API quota `api.jsObjectSize`
- * which limits the number of properties in any JS object to 2000.
- * The quota limit for arrays (`api.jsArraySize`) is 20000, so a total of 40 million products can be stored in changedProducts.
- *
- * The Boolean value of the productID keys indicates whether the product was added/changed (true)
- * or removed with <product mode="delete" product-id="${productID}"/> (false).
- * These products will be retrieved from the database, enriched and sent to Algolia (or marked for deletion).
- *
+ * Returns whether there are any properties in a dense array of objects
+ * @param {Array} objectsArray The array of objects to check, array is filled up sequentially and densely
+ * @returns {boolean} Whether any object in the array contain any properties
+ */
+function isObjectsArrayEmpty(objectsArray) {
+    if (empty(objectsArray)) {
+        return true;
+    }
+
+    // array has at least one element
+    var lastObject = objectsArray[objectsArray.length - 1];
+    if (empty(lastObject)) {
+        return true;
+    }
+
+    return false;
+}
+
+/**
+ * Retrieves the IDs of the products which have changed since the last update
+ * (an attribute of the product's, its inventory levels or its price)
+ * from the supplied XML and adds them to the changedProducts structure.
  * @param {dw.io.File} xmlFile The path to the XML file.
  * @param {Array} changedProducts An array of objects containing the changed products
- * @returns {number|boolean} The number of records read if successful, false if not successful
+ * @param {string} resourceType Type of export file: "catalog" | "inventory" | "pricebook"
+ * @returns {Object} The result object containing `success` (boolean) and `nrProductsRead` (number)
  */
-function updateChangedProductsObjectFromXML(xmlFile, changedProducts) {
+function updateCPObjectFromXML(xmlFile, changedProducts, resourceType) {
     var XMLStreamReader = require('dw/io/XMLStreamReader');
     var XMLStreamConstants = require('dw/io/XMLStreamConstants');
     var FileReader = require('dw/io/FileReader');
     var catalogID;
-    var readRecordsCount = 0;
+    var resultObj = {
+        nrProductsRead: 0,
+        success: false
+    };
 
     try {
         if (xmlFile.exists()) {
             var fileReader = new FileReader(xmlFile);
             var xmlStreamReader = new XMLStreamReader(fileReader);
             var success = false;
 
-            while (xmlStreamReader.hasNext()) {
-                var xmlEvent = xmlStreamReader.next();
+            switch (resourceType) {
+                case 'catalog':
```

Comment:
Suggestion: Since the code is similar for all resource types, we could get rid of the `resourceType` parameter and directly rely on the XML content:

```js
if (['product', 'record', 'price-table'].includes(xmlStreamReader.getLocalName()) {
  var productID = xmlStreamReader.getAttributeValue(null, 'product-id');
  var isAvailable = true;
  if (xmlStreamReader.getLocalName() === 'product') {
    var mode = xmlStreamReader.getAttributeValue(null, 'mode'); // <product mode="delete">
    isAvailable = mode !== 'delete';
  }
  ...
}
```

Reply:
Thought of that, but that's not the best idea due to not all nodes being present from one export XML to the next and the user essentially being able to create new nodes in the XML through custom attributes. E.g. if they create a custom attribute called `record` for the Product system object, the resource type detection is screwed.
This is safer than the alternative as it's 100% guaranteed that the right file is being processed by the right code block.